lib/twom.c

```diff
@@ -0,0 +1,3220 @@
+
+    scratch[79] = 0; // avoid overruns
+
+    printf("UUID: uuid=%s\nFNAME: fname=%s\n", twom_db_uuid(db), twom_db_fname(db));
+    printf("CHECKSUM ENGINE: %s\n", checksum_engine(txn->file));
+    printf("HEADER: v=%lu g=%llu fl=%08lX num=(%llu/%llu) sz=(%08llX/%08llX/%08llX) ml=%lu\n",
+          (LU)header->version,
+          (LLU)header->generation,
+          (LU)header->flags,
+          (LLU)header->num_records,
+          (LLU)header->num_commits,
+          (LLU)header->dirty_size,
+          (LLU)header->current_size,
+          (LLU)header->repack_size,
+          (LU)header->maxlevel);
+
+    while (offset < header->current_size) {
+        printf("%08llX ", (LLU)offset);
+
+        ptr = safeptr(loc, offset);
+        if (!ptr) {
+            printf("ERROR BAD POINTER\n");
+            break;
+        }
+
+        if (check_headcsum(txn, loc->file, ptr, offset)) {
+            printf("ERROR [HEADCSUM %08lX %08lX] ",
+                    (long unsigned) HEADCSUM(ptr),
+                    (long unsigned) loc->file->csum(ptr, HEADLEN(ptr)));
+        }
+
+        if (check_tailcsum(txn, loc->file, ptr, offset)) {
+            printf("ERROR [TAILCSUM %08lX %08lX] ",
+                    (long unsigned) TAILCSUM(ptr),
+                    (long unsigned) loc->file->csum(KEYPTR(ptr), TAILLEN(ptr)));
+        }
+
+        uint8_t type = TYPE(ptr);
+        if (type == COMMIT) {
+            printf("COMMIT start=%08llX\n", (LLU)NEXT0(ptr, 0));
+        }
+        else if (type == DELETE) {
+            printf("DELETE ancestor=%08llX\n", (LLU)NEXT0(ptr, 0));
+        }
+        else {
+            const char *key = KEYPTR(ptr);
+            int len = KEYLEN(ptr);
+            if (len > 79) len = 79;
+            if (key) strncpy(scratch, key, len);
+            scratch[len] = 0;
+            for (i = 0; i < len; i++)
+                if (!scratch[i]) scratch[i] = '-';
+            printf("%s kl=%llu dl=%llu lvl=%d (%s)\n",
+                   typestr[type],
+                   (LLU)KEYLEN(ptr), (LLU)VALLEN(ptr),
+                   LEVEL(ptr), scratch);
+            if (ancestoroffset[type]) {
+                printf("\t%08llX <-\n", (LLU)ANCESTOR(ptr));
+            }
+            printf("\t%08llX %08llX", (LLU)NEXT0(ptr, 0), (LLU)NEXT0(ptr, 1));
+            uint8_t level = LEVEL(ptr);
+            for (i = 1; i < level; i++) {
+                if (!((i-1) % 8))
+                    printf("\n\t");
+                printf("%08llX ", (LLU)NEXTN(ptr, i));
+            }
+            printf("\n");
+            if (detail > 2) {
+                const char *val = VALPTR(ptr);
+                int len = VALLEN(ptr);
+                if (len > 79) len = 79;
+                if (val) strncpy(scratch, val, len);
+                scratch[len] = 0;
+                printf("\tv=(%s)\n", scratch);
+            }
+        }
+
+        offset += RECLEN(ptr);
+    }
+done:
+
+    return r;
+}
+
+int twom_db_fetch(struct twom_db *db,
```

Comment:
Can the public functions in this file be marked as `HIDDEN` ⇔ `__attribute__((__visibility__("hidden")))`, as they are not called from outside libcyrus.so?